### PR TITLE
lib: boost_cpu: introduction of library for clock frequency boosting

### PR DESCRIPTION
**Pull request by awojasinski-nordicsemi**
This commit introduces a new library that enables changing the CPU clock
frequency. It can be done after kernel initialization by setting
`CONFIG_BOOST_CPU_CLOCK_AT_STARTUP` Kconfig symbol or at application
request with introduced API.

Signed-off-by: Adam Wojasinski <adam.wojasinski@nordicsemi.no>

### CODE REVIEW

**Review by pdunaj — APPROVED** [review]

**Comment by greg-fer** [comment]
We're going to need a short documentation page if this lib is to go in.

**Comment by pdunaj** [comment]
Hi @hubertmis , @anangl , my comment here is that it would be beneficial to have the feature exposed instead of hide it in low level api. Also to make it easy to enable from app boot if somebody wants to pay the price of increased power consumption.

This is only for nrf53 but it might be that we might want to have the api like - lock_on_freq in the future.

This said, I am not fully convinced we need it. I would normally say not so much except several people outside already has asked about having clean way of setting this boost.

File: lib/boost_cpu/boost_cpu.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include <zephyr.h>
+#include <nrfx_clock.h>
+
+#if CONFIG_BOOST_CPU_CLOCK_AT_STARTUP
+#include <init.h>
+#endif
+
+LOG_MODULE_REGISTER(boost_cpu, CONFIG_BOOST_CPU_LOG_LEVEL);
+
+void cpu_boost(bool enable)
```
Comment (hubertmis) [inline lib/boost_cpu/boost_cpu.c:17]: How about `enum` or `int` to potentially reuse this API for CPUs with more than two operationg points?

**Review by hubertmis — COMMENTED** [review]
@pdunaj, I think library like this should remain a low level driver (potentially Zephyr driver in the future). We should create a service in power management subsystem responsible for aggregating on-the-fly CPU frequency requirements from multiple software modules (similarly we already aggregate latency requirements), and this power management aggregator should have exposed API, and use the low level driver created in this patch.

**Comment by greg-fer** [comment]
> I think library like this should remain a low level driver

Regardless of what you decide, @pdunaj , @hubertmis , @anangl , if this is going to be a library or a driver, it needs documentation. :)

**Comment by greg-fer** [comment]
@awojasinski-nordicsemi , @pdunaj , a draft of the library docs: https://github.com/nrfconnect/sdk-nrf/pull/7608

**Comment by anangl** [comment]
> Hi @hubertmis , @anangl , my comment here is that it would be beneficial to have the feature exposed instead of hide it in low level api. Also to make it easy to enable from app boot if somebody wants to pay the price of increased power consumption.
> 
> This is only for nrf53 but it might be that we might want to have the api like - lock_on_freq in the future.
> 
> This said, I am not fully convinced we need it. I would normally say not so much except several people outside already has asked about having clean way of setting this boost.

In my opinion such piece of code does not really deserve to be called a library. It is available only for one SoC and it just wraps a single call to an nrfx function in a bit simpler API. The added Kconfig option is also not very flexible, as one might want to use the higher frequency also during kernel initialization.
I think that for people asking how to change CPU frequency on nRF5340, a sample application that just presents how to do it would be a better thing.
I agree that a direct call to `nrfx_clock` API does not look like a clean solution, so if we'd want to provide something that looks better, I think we should rather extend the [nRF-specific clock_control API](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/clock_control/nrf_clock_control.h) in Zephyr instead.

**Comment by awojasinski-nordicsemi** [comment]
> > Hi @hubertmis , @anangl , my comment here is that it would be beneficial to have the feature exposed instead of hide it in low level api. Also to make it easy to enable from app boot if somebody wants to pay the price of increased power consumption.
> > This is only for nrf53 but it might be that we might want to have the api like - lock_on_freq in the future.
> > This said, I am not fully convinced we need it. I would normally say not so much except several people outside already has asked about having clean way of setting this boost.
> 
> In my opinion such piece of code does not really deserve to be called a library. It is available only for one SoC and it just wraps a single call to an nrfx function in a bit simpler API. The added Kconfig option is also not very flexible, as one might want to use the higher frequency also during kernel initialization. I think that for people asking how to change CPU frequency on nRF5340, a sample application that just presents how to do it would be a better thing. I agree that a direct call to `nrfx_clock` API does not look like a clean solution, so if we'd want to provide something that looks better, I think we should rather extend the [nRF-specific clock_control API](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/clock_control/nrf_clock_control.h) in Zephyr instead.

Another solution that solves my needs is prepared in this commit https://github.com/awojasinski-nordicsemi/sdk-nrf/commit/ead84303e68c54daa26ffca2b886b2c5f9d82fa1

**Comment by awojasinski-nordicsemi** [comment]
After some discussion with @pdunaj we decided not to close this branch. This feature will not have a separate library.

**Pull request closed without merging**